validator_client/src/attestation_service.rs

```diff
@@ -137,19 +137,12 @@ impl<T: SlotClock + 'static, E: EthSpec> AttestationService<T, E> {
             "next_update_millis" => duration_to_next_slot.as_millis()
         );
 
-        let mut interval = {
-            // Note: `interval_at` panics if `slot_duration` is 0
-            interval_at(
-                Instant::now() + duration_to_next_slot + slot_duration / 3,
-                slot_duration,
-            )
-        };
-
         let executor = self.context.executor.clone();
 
         let interval_fut = async move {
-            loop {
-                interval.tick().await;
+            while let Some(duration_to_next_slot) = self.slot_clock.duration_to_next_slot() {
```

Comment:
If we happen to be unable to read the system clock then this timer is going to exit and never run again. I'm not sure why that would happen, but it seems nice to be resilient against it.

I took an approach here where I just sleep for a slot if we can't read the system clock, that gives us a chance to recover if the slot clock comes good again:

https://github.com/sigp/lighthouse/blob/46920a84e857d6c086adf4b42810ac03de94ed3e/beacon_node/beacon_chain/src/state_advance_timer.rs#L99-L107

Reply:
Nice idea. Fixed this.